sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -467,10 +494,148 @@ The `node` class provides the {crs}[common reference semantics].
+representation of the argument. The raw byte representation is intended to
+enable updating arguments set using
+link:../proposed/sycl_ext_oneapi_raw_kernel_arg.asciidoc[sycl_ext_oneapi_raw_kernel_arg].
+
+Dynamic parameters are registered with nodes in a modifiable graph, with each
+registration associating one or more node arguments to the dynamic parameter
+instance. Registration happens inside the command-group that the node represents
+, and is done using the argument index (known when argument originally set using
+`set_arg()`/`set_args()`). It is valid for a node argument to be registered with
+more than one dynamic parameter instance.
+
+Registration is done by using the argument index (known when argument
+originally set using `set_arg()`/`set_args()`).
+
+See <<executable-graph-update, Executable Graph Update>> for more information
+about updating node parameters.
+
+The `dynamic_parameter` class provides the {crs}[common reference semantics].
+
+Table {counter: tableNumber}. Member functions of the `dynamic_parameter` class.
+[cols="2a,a"]
+|===
+|Member Function|Description
+
+|
+[source,c++]
+----
+dynamic_parameter(command_graph<graph_state::modifiable> graph);
+----
+|Constructs a dynamic parameter object that can be registered with node
+arguments.
+
+Parameters:
+
+* `graph` - Graph containing the nodes which will have dynamic parameters.
+
+|
+[source,c++]
+----
+dynamic_parameter(const ValueT &initialValue, command_graph<graph_state::modifiable> graph);
+----
+|Constructs a dynamic parameter object that can be registered with node
+arguments, while providing an intiial value.
+
+Parameters:
+
+* `initialValue` - Initial value for the dynamic parameter.
+* `graph` - Graph containing the nodes which will have dynamic parameters.
+
+|
+[source,c++]
+----
+void register(int argIndex, const handler &cgh);
+----
+|Callable only from command-group scope when adding a node to a graph
+using the explicit API. Associate the dynamic parameter
+with the kernel parameter of the node that encapsulates this command-group.
+Parameters are associated by their index which was given when they were
+registered with `handler`.
+
+Parameters:
+
+* `argIndex` - Kernel argument index starting from zero to associate with
+  the dynamic parameter.
+* `handler` - SYCL handler object.
+
+Exceptions:
+
+* Throws with error code `invalid` if `argIndex` is not a valid argument index
+  for the resulting node.
+* Throws with error code `invalid` if the type of the command-group is not a
+  kernel execution.
+* Throws with error code `invalid` if the resulting node is not associated with
+  the graph that this dynamic parameter was created with.
+
+|
+[source,c++]
+----
+void update(const ValueT& newValue);
+----
+
+|Updates parameters in all nodes registered with this dynamic
+parameter to `newValue`. This new value will be reflected immediately in the
+modifiable graph which contains the registered nodes. It is not an
```

Comment:
That is fair I think this description is slightly incorrect. In reality the flow is like:
- `dynamic_parameter::update()` updates the values inside the registered `node`s immediately as I previously described.
- `command_graph::update()` the equivalent internal nodes  (we don't expose these to the user) and backend structures (command-buffers) to those which are passed to the function, are updated to reflect the current state of the nodes passed to the function. This is where the actual kernel arguments are updated in the backend.

I'll update the wording of `command_graph::update()` to correctly reflect this.

Reply:
Oh, I just realized that the `node` represents a node in the modifiable graph, which is different from the executable graph.  Now, I'm wondering what happens in a case like this:

```
sycl_ext::command_graph myGraph(myContext, myDevice);  // Creates a modifiable graph

int myScalar = 42;
dynamic_parameter dynParamScalar(myGraph, myScalar);

node nodeA = myGraph.add([&](handler& cgh) {
  dynParamScalar.register_with_node(cgh, 0);
  cgh.parallel_for(range {n}, myKernel);
});

dynParamScalar.update(1024);  // Change the value of the parameter before the graph is finalized

auto execGraph = myGraph.finalize();
```

What is the value of the `myScalar` parameter when the graph is executed?  Is it 42 or 1024?

Reply:
> What is the value of the myScalar parameter when the graph is executed? Is it 42 or 1024?

The value in this case would be 1024. An executable graph is basically a snapshot of the modifiable graph fixed at the point where it was created. Since the parameter was updated before `execGraph` was created through calling`finalize()` then it will contain the updated arg value.

Reply:
In that case, I would recommend changing this sentence:

> This new value will be reflected immediately in the modifiable graph which contains the registered nodes.

with something like:

> This new value will be reflected when the modifiable graph is finalized (by calling `command_graph::finalize`) or when an executable graph containing the registered node is updated (by calling `command_graph::update` and passing the relevant node).

Reply:
I feel like the suggested wording is a bit confusing since it makes it seem like nothing will actually happen until those two things are done. I've attempted a combination of both which hopefully is clearer though:

> This new value will be reflected immediately in the modifiable graph
which contains the registered nodes. The new value will not be reflected in any
executable graphs created from that modifiable graph until
`command_graph::update()` is called passing the modified nodes, or a new
executable graph is finalized from the modifiable graph.